### PR TITLE
Tens ubuntu 22.04 migration

### PR DESCRIPTION
**Pull request by priyadarshitathagat**
The PR :- 

[1] Migrates TE-Controller and TE-DP containers from 16.04 to 22.04 Ubunbtu.
[2] Support for IPV6 UDP traffic has been added.
[3] Deprecated `setup.sh` script which was earlier used to build and install custom packages.
[4] Fixed CVE :- https://github.com/vmware/te-ns/security/dependabot/9

### CODE REVIEW

File: te/TE_DP_CONFIG.py [hunk 1]
```diff
@@ -118,13 +118,11 @@ def __del__(self):
             te_r_queue.empty()
         del self.__host_ip
         del self.__STATES
-        del self.__host_ip
```
Comment (priyadarshitathagat) [inline te/TE_DP_CONFIG.py:121]: trying to del already deleted instance var

File: te/TE_UTILS.py [hunk 2]
```diff
@@ -41,7 +41,7 @@
 
 from threading import Thread
 import logging, logging.handlers
-from collections import Mapping
+from collections.abc import Mapping
```
Comment (priyadarshitathagat) [inline te/TE_UTILS.py:44]: the class has been moved to a different lib.

File: te/te_docker/Dockerfile [hunk 3]
```diff
@@ -118,7 +120,7 @@ RUN chmod -R 755 /var/www/html/
 RUN echo "Done TE Dockerization!\n"
 
 ENTRYPOINT sed -i "s/5001/$NGINX_PORT/g" /etc/nginx/sites-available/default && \
-    redis-server --port $REDIS_PORT --daemonize yes && /etc/init.d/nginx restart && \
+    redis-server --port $REDIS_PORT --protected-mode no --daemonize yes && /etc/init.d/nginx restart && \
```
Comment (priyadarshitathagat) [inline te/te_docker/Dockerfile:123]: the default value of protected mode has changed in the latest ubuntu version, causing failure to connect to te_dp machines.

File: te/te_docker/Dockerfile [hunk 4]
```diff
@@ -75,12 +75,14 @@ missingok \n\
...
 
 #Creating a cron (Running cron job every 2 mins -- logrotate)
+RUN apt-get install -y cron
 RUN echo "*/2 * * * * /usr/sbin/logrotate /etc/logrotate.d/te-logs" > /etc/cron.d/cron_logrotate_te_logs
 RUN chmod 0644 /etc/cron.d/cron_logrotate_te_logs
 RUN crontab /etc/cron.d/cron_logrotate_te_logs
 
 COPY te_docker/requirements.txt /tmp/
-RUN pip3 install --upgrade "pip < 21.0"
+RUN apt install libpq-dev -y
```
Comment (priyadarshitathagat) [inline te/te_docker/Dockerfile:84]: needed for postgres.

File: te/tedp_docker/Dockerfile [hunk 5]
```diff
@@ -112,10 +112,8 @@ ADD te/TE_UTILS.py \
 RUN mkdir -pv $WORKDR/bin
 COPY --from=tedp_bin:v2.0 $WORKDR/bin $WORKDR/bin
 COPY --from=tedp_bin:v2.0 $WORKDR/usr_lib64_deps.tar.gz ${usr_lib64_path}/usr_lib64_deps.tar.gz
-COPY --from=tedp_bin:v2.0 $WORKDR/usr_lib_deps.tar.gz ${usr_lib_path}/usr_lib_deps.tar.gz
 COPY --from=tedp_bin:v2.0 $WORKDR/lib64_deps.tar.gz ${lib64_path}/lib64_deps.tar.gz
-RUN tar -xvf ${usr_lib_path}/usr_lib_deps.tar.gz -C / && rm ${usr_lib_path}/usr_lib_deps.tar.gz && \
-    tar -xvf ${usr_lib64_path}/usr_lib64_deps.tar.gz -C / && rm ${usr_lib64_path}/usr_lib64_deps.tar.gz && \
+RUN tar -xvf ${usr_lib64_path}/usr_lib64_deps.tar.gz -C / && rm ${usr_lib64_path}/usr_lib64_deps.tar.gz && \
```
Comment (priyadarshitathagat) [inline te/tedp_docker/Dockerfile:116]: Use of setup.sh has been deprecated and hence no custom packages has been installed.

File: te/tedp_docker/requirements.txt [hunk 6]
```diff
@@ -29,10 +29,10 @@
 # OF SUCH DAMAGE
 #**********************************************************************************************
 
-greenlet==0.4.15
-redis==3.1.0
+greenlet==2.0.2
+redis==4.5.3
 requests==2.25.1
-rq==0.13.0
+rq==1.11.1
```
Comment (priyadarshitathagat) [inline te/tedp_docker/requirements.txt:35]: upgraded to latest versions

File: te_dp/Makefile [hunk 7]
```diff
@@ -39,20 +39,20 @@ BINDIR = bin
...
 
 CFLAGS     = -g -O0 -Wall -I.
-STAT_FLAGS = -pthread -lzmq -ljson
+STAT_FLAGS = -pthread -lzmq -ljson-c
 
 all: dp stat
 
 dp: $(DP_OBJ)
-	@$(CC) $(CFLAGS) -o $(DP_BIN) $^ $(DP_FLAGS)
+	@$(CC) $(CFLAGS) -o $(DP_BIN) $^ $(DP_FLAGS) -no-pie
```
Comment (priyadarshitathagat) [inline te_dp/Makefile:55]: Issue seen :- without `-no-pie`  
https://stackoverflow.com/questions/73275183/gcc-assembly-x86-relocation-warning-creating-dt-textrel-in-a-pie

Further reference:- 
https://stackoverflow.com/a/50615370 -> read comments as well

File: te_dp/src/te_stats_collector.cpp [hunk 8]
```diff
@@ -991,8 +988,6 @@ void stat_collector_thread(const char* tcp_socket_to_connect, vector<int> &activ
 
 cleanup:
     json_object_put(metrics);
-    for (auto it=metrics_obj_to_clear.begin(); it!=metrics_obj_to_clear.end(); it++)
-        json_object_put(*it);
```
Comment (priyadarshitathagat) [inline te_dp/src/te_stats_collector.cpp:995]: We were double freeing and that was leading to `malloc_consolidate` crash in te-stats-collector thereby killing te_dp process. The upgrade of json-c library may be the culprit,  the code  causing the issue was added as part of mem-leak issue, but looks like the new version of `json_object_object_add` frees the object that was used as val which need not be freed later
```
JSON_EXPORT int json_object_object_add	(	struct json_object * 	obj,
const char * 	key,
struct json_object * 	val 
)		
Add an object field to a json_object of type json_type_object

The reference count of val will not be incremented, in effect transferring ownership that object to obj, and thus val will be freed when obj is. (i.e. through json_object_put(obj))
```
https://json-c.github.io/json-c/json-c-current-release/doc/html/json__object_8h.html#a27bd808a022251059a43f1f6370441cd

**Comment by priyadarshitathagat** [comment]
@aravindhank11 -  Please review, I am not able to add you as a reviewer.

**Comment by aravindhank11** [comment]
Sure @priyadarshitathagat. I'll need some time though ~3-4 days.

**Comment by priyadarshitathagat** [comment]
@aravindhank11 - gentle reminder on the review.

**Comment by aravindhank11** [comment]
Sorry! I'll take a pass postively by the weekend

**Comment by aravindhank11** [comment]
Thank you for the changes, @priyadarshitathagat. Please get it reviewed from @vipinpr5991 as well, please!

File: te/te_docker/Dockerfile (hunk 4 above)
Comment (priyadarshitathagat) [reply te/te_docker/Dockerfile:84]: The dev package contains minimal set of PostgreSQL binaries and headers, it should be fine. Also, it provides all requirements of Psycopg2, and I faced pg_config error during the build which was fixed by the dev package.
https://stackoverflow.com/questions/35104097/how-to-install-psycopg2-with-pg-config-error

**Review by vipinpr5991 — APPROVED** [review]
LGTM, thanks for the changes @priyadarshitathagat

**Pull request merged**